feat!: add AztecAddress.isValid and make random be valid

nventuro opened the pull request:
Closes https://github.com/AztecProtocol/aztec-packages/issues/10039

Since https://github.com/AztecProtocol/aztec-packages/issues/8970 is not yet implemented, invalid addresses cannot be the recipient of e.g. token transfers, since we end up producing points not in the curve and causing MSM to fail. Making `random()` return valid addresses seems like very reasonable behavior.

nventuro committed:
Add isValid and change random

nventuro committed:
Fix formatting

iAmMichaelConnor commented on yarn-project/foundation/src/aztec-address/index.ts:
```diff
@@ -69,8 +69,17 @@ export class AztecAddress {
     return new AztecAddress(hexToBuffer(buf));
   }
 
+  /**
```
Also (I can't comment on unchanged code), shouldn't the constructor of `Address` be ensuring the `value` passed in is a valid x-coordinate? And perhaps the name of the property `value` should be `x` (or something), to hint at the fact this value gets converted into a point.
Also also, shouldn't this `Address` class have a `toAddressPoint` method?
The doc comment at the top of the `Address` class should also provide more details relating to points.

iAmMichaelConnor replied:
Also also, there's a `zero()` method, but is `0` a valid x-coordinate? Does -17 have a square root in the field?

iAmMichaelConnor replied:
> And it's also fine to use invalid addresses
Is it?

I just worry that if the constructor doesn't enforce that the address is valid, then a user might create an invalid address and send funds to it. (Sure, the contract facilitating the transfer should have constrains to prevent this, but defense in depth, perhaps?).
Maybe Aztec's "zero address" should be `5` (or whatever the smallest valid x is)?

nventuro committed:
Add clarifying comments

nventuro committed:
Add random test

nventuro committed:
Add Address.ToAddressPoint

nventuro commented:
I refactored `Point` a little bit to reuse this `YFromX` method, and added `AztecAddress.toAddressPoint`, which simplified some other parts of the encryption flow.

nventuro committed:
Merge branch 'master' into nv/valid-addr

nventuro committed:
Fix unused import

nventuro committed:
Fix formatting
